### PR TITLE
Custom easyocr

### PR DESCRIPTION
**Pull request by siddhantcarscan**
1. Added seperate paths for separate files. such as .yaml, .py, .pth
2. modified device = device. instead of gpu= bool.
3. Modified loading of model from single model using key. ['easyocr_weight']

### CODE REVIEW

File: easyocr/easyocr.py [hunk 1]
```diff
@@ -423,8 +444,522 @@ def recognize(self, img_cv_grey, horizontal_list=None, free_list=None,\
...
+
+
+
+
+class Custom_Reader(object):
+
+    def __init__(self, lang_list, opt=None, device=None, model_storage_directory=None,
+                 user_network_directory=None, detect_network="craft", download_enabled=True, 
+                 detector=True, recognizer=True, verbose=True, 
+                 quantize=True, cudnn_benchmark=False, model_path=None ,yaml_path=None, model_arch=None):
```
Comment (PushpakBhoge512) [inline easyocr/easyocr.py]: we have added `model_path=None ,yaml_path=None, model_arch=None)` remove this parameters

File: easyocr/easyocr.py [hunk 2]
```diff
@@ -423,8 +444,522 @@ def recognize(self, img_cv_grey, horizontal_list=None, free_list=None,\
...
+        self.model_storage_directory = MODULE_PATH + '/model'
+        if model_storage_directory:
+            self.model_storage_directory = model_storage_directory
+        Path(self.model_storage_directory).mkdir(parents=True, exist_ok=True)
+
+        self.user_network_directory = MODULE_PATH + '/user_network'
+        if user_network_directory:
+            self.user_network_directory = user_network_directory
+        Path(self.user_network_directory).mkdir(parents=True, exist_ok=True)
+        sys.path.append(self.user_network_directory)
```
Comment (PushpakBhoge512) [inline easyocr/easyocr.py]: remove all of this directly set
```
self.model_path=model_path
self.yaml_path=yaml_path
self.model_arch=model_arch
```

File: easyocr/easyocr.py [hunk 3]
```diff
@@ -423,8 +444,522 @@ def recognize(self, img_cv_grey, horizontal_list=None, free_list=None,\
...
+
+        self.detection_models = detection_models
+        self.recognition_models = recognition_models
+
+        # check and download detection model
+        self.support_detection_network = ['craft', 'dbnet18']
+        self.quantize=quantize, 
+        self.cudnn_benchmark=cudnn_benchmark
+        if detector:
+            detector_path = self.getDetectorPath(detect_network)
```
Comment (PushpakBhoge512) [inline easyocr/easyocr.py]: delete everythinhg related to detector where ever possible

File: easyocr/easyocr.py [hunk 4]
```diff
@@ -423,8 +444,522 @@ def recognize(self, img_cv_grey, horizontal_list=None, free_list=None,\
...
+                    model = recognition_models['gen2']['cyrillic_g2']
+                    recog_network = 'generation2'
+                else:
+                    self.model_lang = 'latin'
+                    model = recognition_models['gen2']['latin_g2']
+                    recog_network = 'generation2'
+            self.character = model['characters']
+
+            # model_path = os.path.join(self.model_storage_directory, model['filename'])
+            # check recognition model file
```
Comment (PushpakBhoge512) [inline easyocr/easyocr.py]: delete this and raise exception with message "pretrained model loading not supported"

File: easyocr/easyocr.py [hunk 5]
```diff
@@ -423,8 +444,522 @@ def recognize(self, img_cv_grey, horizontal_list=None, free_list=None,\
...
+            # model_file = recog_network+ '.pth'
+            # model_path = 
+            self.setLanguageList(lang_list, recog_config)
+
+        dict_list = {}
+        for lang in lang_list:
+            dict_list[lang] = os.path.join(BASE_PATH, 'dict', lang + ".txt")
+
+        if detector:
+            self.detector = self.initDetector(detector_path)
```
Comment (PushpakBhoge512) [inline easyocr/easyocr.py]: delete this

File: easyocr/easyocr.py [hunk 6]
```diff
@@ -423,8 +444,522 @@ def recognize(self, img_cv_grey, horizontal_list=None, free_list=None,\
...
+                LOGGER.warning(corrupt_msg)
+                os.remove(detector_path)
+                LOGGER.warning('Re-downloading the detection model, please wait. '
+                               'This may take several minutes depending upon your network connection.')
+                download_and_unzip(self.detection_models[self.detect_network]['url'], self.detection_models[self.detect_network]['filename'], self.model_storage_directory, self.verbose)
+                assert calculate_md5(detector_path) == self.detection_models[self.detect_network]['md5sum'], corrupt_msg
+        else:
+            raise RuntimeError("Unsupport detector network. Support networks are {}.".format(', '.join(self.support_detection_network)))
+        
+        return detector_path
```
Comment (PushpakBhoge512) [inline easyocr/easyocr.py]: delete

File: easyocr/easyocr.py [hunk 7]
```diff
@@ -423,8 +444,522 @@ def recognize(self, img_cv_grey, horizontal_list=None, free_list=None,\
...
+    def initDetector(self, detector_path):
+        return self.get_detector(detector_path, 
+                                 device = self.device, 
+                                 quantize = self.quantize, 
+                                 cudnn_benchmark = self.cudnn_benchmark
+                                 )
+    
+    def setDetector(self, detect_network):
+        detector_path = self.getDetectorPath(detect_network)
+        self.detector = self.initDetector(detector_path)
```
Comment (PushpakBhoge512) [inline easyocr/easyocr.py]: delete

File: easyocr/easyocr.py [hunk 8]
```diff
@@ -423,8 +444,522 @@ def recognize(self, img_cv_grey, horizontal_list=None, free_list=None,\
...
+                                                        (optimal_num_chars is None))
+            if min_size:
+                horizontal_list = [i for i in horizontal_list if max(
+                    i[1] - i[0], i[3] - i[2]) > min_size]
+                free_list = [i for i in free_list if max(
+                    diff([c[0] for c in i]), diff([c[1] for c in i])) > min_size]
+            horizontal_list_agg.append(horizontal_list)
+            free_list_agg.append(free_list)
+
+        return horizontal_list_agg, free_list_agg
```
Comment (PushpakBhoge512) [inline easyocr/easyocr.py]: delete

**Review by PushpakBhoge512 — CHANGES_REQUESTED** [review]

**Comment by siddhantcarscan** [comment]
Removed the unnecessary functions and params from Custom_Reader.

File: easyocr/easyocr.py [hunk 9]
```diff
@@ -568,3 +587,191 @@ def readtext_batched(self, image, n_width=None, n_height=None,\
...
+                
+            available_lang = recog_config['lang_list']
+            self.setModelLanguage(recog_network, lang_list, available_lang, str(available_lang))
+            #char_file = os.path.join(self.user_network_directory, recog_network+ '.txt')
+            self.character = recog_config['character_list']
+            # model_file = recog_network+ '.pth'
+            # model_path = 
+            self.setLanguageList(lang_list, recog_config)
+        except:
+            print('user-defined model is not provided')
```
Comment (PushpakBhoge512) [inline easyocr/easyocr.py]: you have to raise error here not just print statement

**Review by PushpakBhoge512 — CHANGES_REQUESTED** [review]

**Review by PushpakBhoge512 — APPROVED** [review]

**Pull request merged**